### ISSUE TITLE
Tuple

### ISSUE THREAD

**Issue by igorbonadio** [positron-lang/spec#15]
Let's talk about tuples.

**Comment by igorbonadio** [comment]
I think tuples should be one of the most important kind of objects. For example:
- `nil` and `()`(the empty tuple) are equal and both are of type `Nil` or `()`
- any expressions is equal to a tuple of 1 element. So both are of `type_of(expression)`
- tuples of n > 1 elements are of type `(Type1, Type2, ...)`, where Type1, Type2 can be any type (equal or different)

And here I think we have a good example of pattern matching (issue #16 ):

``` ruby
let (x, y, z) = (1, 2, 3)
# parentheses are optional
let x, y, z = 1, 2, 3
```

**Comment by renatocf** [comment]
Marking as opened to #18

**Comment by igorbonadio** [comment]
As we discussed in #13 , I'll propose:

## Tuples
- `()` is the empty tuple and its type is `()`
- any expressions is equal to a tuple of 1 element. So both are of `type_of(expression)`
- tuples of n > 1 elements are of type `(Type1, Type2, ...)`, where `Type1`, `Type2` can be any type (equal or different)

Grammar:

``` antlr
tuple : '(' tuple_elements? ')'
      | tuple_elements
      ;

tuple_elements : (expression ',')* expression
               ;
```

Examples:

``` ruby
(1, 2, 3.3) # tuple of type (Int, Int, Real)
1, 2, 3.3   # parentheses are optional
```

**Comment by vinivendra** [comment]
Now that we're almost done discussing `nil`s, do you think empty tuples `()` should be equivalent to `nil`? It makes sense to me semantically. For instance, when you think of functions, a function with no return value should return "nothing", no matter if we choose to call nothing `nil` or `()`; they represent the same general idea.

**Comment by igorbonadio** [comment]
For example, I think the answer for the following expressions are:

``` ruby
def function() -> ()
end

def function2() -> (tuple: (Int, Int))
  tuple = nil
end

nil == ()         # false, because () is an instance of (), so it is initialized... and nil means uninitialized
function() == ()  # true
function() == nil # false

function2() == ()  # false as type (Int, Int) is different of ()
function2() == nil # true as the type (Int, Int) is uninitialized
```

What do you think?

**Comment by vinivendra** [comment]
I don't know. I mean, I understand what you proposed, but what's to be gained from having an empty tuple which is different than our "nothing" value?

To me, the definition

``` ruby
def function() -> ()
end
```

means that a function doesn't receive anything as an argument and doesn't return anything to its user. That translates to returning `nil`, in my opinion. I just don't see the point of allowing another representation of this return value to exist (even if it is meant as a different element), I think it would just be confusing.

**Comment by igorbonadio** [comment]
Yes, It is simpler.

But, I was thinking if it will not be confusing. For example:

``` ruby
def function() -> ()
end

def function2() -> (tuple: (Int, Int))
  tuple = nil
end

function() == function2() # it should be true or false?
```

**Comment by renatocf** [comment]
For me, this example should be an error!

If we analyze our definition of `nil`, we defined it as a special type that can be assigned in variables / values of any type. However, an equality method as above should accept only a caller and a callee of the same type. As `()` is the empty tuple and `(tuple: (Int, Int))` is a tuple of a pair, they cannot be compared.

Here, it makes sense that `()` equals `nil` because, as a tuple, `()` is an object, and therefore can be assigned to `nil`. However, as it is an empty tuple, it cannot have any other value, which leads us that `()` is always `nil`. On the other hand, `(tuple : (Int, Int))` may be `nil` or any assignment of values allowed by the elements of the tuple.

**Comment by vinivendra** [comment]
I agree with most of what @renatocf said, but I don't think it should be an error. In fact, I think it should probably be `true`.

In the first example, we aren't actually returning anything from inside the function. I mean, there is no "`result = ()`" to let us know that the returning value will be an `()`, nothing other from the fact that that's the type we're asking for. That's one of the reasons why I think the first function should really return `nil`.

As for the result, we come to a question I think we've already discussed once (though I can't remember its conclusion). Though `==` is meant to work as a function, we might have to allow for a special case here, since doing `nil == blah` would mean sending a message to `nil` (and, therefore, always triggering an error). Either that or we could create a symbolic `Nil` class to implement this behavior:

``` ruby
class Nil
    function ==(rhs: Any) -> (result: Bool)
        result = (rhs == nil) # This wouldn't work, but you get the idea.
    end
end
```

We could trigger a special case or create a new class. Either way, since I'm assuming as above that the first function returns `nil`, it seems to me that the result of that equality should be `true`.

**Comment by igorbonadio** [comment]
Hey guys, I was a little bit busy during last days...

I though about `()` and `nil` and I think it is ok that `() == nil` (sending a message `==` to `()` with `nil` as argument evaluates to `true`).

But the other side (`nil == ()`) is strange... If `nil` is of type `Nil`, then we have a type check problem:

For example,  `var x: Int = nil` is invalid... as x is of type `Int` and `nil` is of type `Nil`.

**Comment by Kazuo256** [comment]
In C++11, the `nullptr` pointer literal has type `nullptr_t` and is implicitly convertible to any pointer type (but not numeric valies, avoid issues with polymorphism). Would a similar approach be of help here?

**Comment by igorbonadio** [comment]
Maybe the simplest solution is, in this case, allow an implicit type cast:

``` ruby
var x: Int = nil
# is the same as
var x: Int = nil as Int
```

And I'm supposing that the class hierarchy is:

```
-----------
| Nil     |
-----------
    ^
    |
-----------
| Object |
-----------
    ^
    |
-----------
| Int     |
-----------
```

So when I send a message to `x`:

``` ruby
x.do_something()
```

Positron sends `do_something()` to an `Integer`. But, during runtime, it is `nil` that doesn't understand it an so the program breaks.

But, if I send a message that `nil` understands, the program works:

``` ruby
x == 1 # => false
```

What do you think? Are there other solutions?

**Comment by Kazuo256** [comment]
Making every method call use dynamic dispatch might cause performance issues, infringing the third law.

**Comment by igorbonadio** [comment]
It is not a dynamic dispatch... For example, imagine that the table of methods of each class is defined as:

| Nil |
| --- |
| == |

| Int |
| --- |
| == |
| do_something |

When I call `x.do_something()` and `x = nil`, then the program will try to access the second function pointer... and it will crash the program.

But when I call `==` everything goes fine (as both have the first function pointer).

**Comment by Kazuo256** [comment]
But, in order to determine which `==` to use by inheritance at runtime, dynamic dispatch would be used, wouldn't it?

**Comment by igorbonadio** [comment]
Yes, in fact it is a kind of dynamic dispatch. When you talked about it, I was thinking in something like java or any dynamic language... languages that really send messages.

It is like a `virtual` function in c++. But I don't know how to do polymorphism without it...

And we can implement it with only one indirection, as I described later. And for some operators we can optimize it (for example `+`, `-`, etc).

It would be something like:

``` c++
obj == obj
// compiles to
obj->klass->equals(obj, obj2)); // c++
```

I dont think we are losing so much...

**Comment by Kazuo256** [comment]
Dynamic dispatch _per se_ isn't that slow. The problem is if you use it too much or your hierarchy trees are too complex.

But I may be indeed worrying too much. Besides you suggestion for optimization, there seems to be a few caching techniques that could avoid traversing whole inheritance trees all the time.

Then - going back to the `==` issue - your proposal shoud work. The only thing that bothers me in that case is that everything would inherit from `Nil` (if I understood correctly), generating some peculiar behaviours like any value matching to `Nil` in pattern matchings. Is that ok?

**Comment by igorbonadio** [comment]
I don't see why it will get slow when the hierarchy grows... Every call will be only one indirection. Isn't it? I think we don't have to transverse the whole inheritance tree...

``` c++
// c++ code
self->klass->method(self, params); // every call can be like this one
```

It will only be a problem in interfaces/traits (I think...).

**Comment by igorbonadio** [comment]
`Nil` can be separated from the main hierarchy. We have to think more about it.

But I don't think it will be a problem if everything inherit from `Nil`. For example:

``` ruby
class Nil # as @vinivendra suggested
  def ==(rhs: Any) -> (result: Bool)
    result = (rhs == nil) # This wouldn't work, but you get the idea.
  end
end

match (123, "abc")
  case (nil, "abc") => false # will not match as `nil != 123`
  case (123, "abc") => true # will match as `123 == 123 and "abc" == "abc"
end
```

The only strange thing would be something like:

``` ruby
var x: Nil = 123
```

**Comment by Kazuo256** [comment]
The approach

``` c++
self->klass->method(self, params);
```

Will have one indirection only if the immediate class of the object has an implementation for the method. Otherwise, it will have to delegate to its parent class. It is in these cases that the long hierarchy makes it slow. You could have a local cache of parent methods in the child classes as I said, though. It would just have to be a smart enough implementation not to let too much redundancy waste memory away.

**Comment by Kazuo256** [comment]
About the pattern matchings with `Nil`, would it be possible to match to a value's type? Like this:

``` ruby
var x: Nil = "string";
match (x) # with a single value to simplify
  case (x is Nil) => 0
  case (x is Int) => x + x
  case (x is String) => x as Int
end
```

If it is, then in this case, no matter what value `x` has, it would always be matched to the first case, I believe (assuming the cases are checked in order).

**Comment by igorbonadio** [comment]
But at compile time we can set the correct pointer to each method.

For example:

``` ruby
class A
  def m1()
  end
  def m2()
  end

  def m3()
  end
end

class B < A
  def m2()
  end
end

class C < B
  def m1()
  end
end
```

At compile time we can generate the folowing data structures:

| address | functions |
| --- | --- |
| 0 | A::m1 |
| 1 | A::m2 |
| 2 | A::m3 |
| 3 | B::m3 |
| 4 | C::m1 |

| A | address |
| --- | --- |
| m1 | 0 |
| m2 | 1 |
| m3 | 2 |

| B | address |
| --- | --- |
| m1 | 0 |
| m2 | 3 |
| m3 | 2 |

| C | address |
| --- | --- |
| m1 | 4 |
| m2 | 3 |
| m3 | 2 |

In this case:

``` ruby
c = C.new
c.m3()
```

`c.m3()` will call the second function, which is A::m3. I think we just need one indirection.

``` ruby
A::m3(c) # I'm supposing that the first parameter of this function is `self`
```

And I'm supposing that we can't change classes during runtime as it is possible in ruby and smalltalk.

**Comment by igorbonadio** [comment]
Now about patter matching. Yes. You are right. But I think it is the right behavior... it is similar to:

``` ruby
var x: Object = "string";
match (x) # with a single value to simplify
  case (x is Object) => 0
  case (x is Int) => x + x
  case (x is String) => x as Int
end
```

If you want to check if something is nil, you should do:

``` ruby
var x: Object = "string";
match (x) # with a single value to simplify
  case (x == nil) => 0
  case (x is Int) => x + x
  case (x is String) => x as Int
end
```

And we are supposing that exist a type check operator (`is`).

**Comment by igorbonadio** [comment]
And we can force user to not create Nil instances. For example:

``` ruby
var x: Nil = 123 # error
```

**Comment by Kazuo256** [comment]
Regarding dynamic dispatch, if you copy all method references from parent classes to child classes, you are spending extra memory (since pointers are also stored in memory) in order to improve efficency. That is what I meant by caching, except that with a dedicated caching algorithm you could have control over the trade-off between memory and performance. If we assume that solution is not possible because the classes' vtables become fixed after compile-time, then we should at least be aware that keeping all this redundancy of method addresses comes at a (perhaps small?) cost.

**Comment by Kazuo256** [comment]
And about pattern matches, if that is the desired behaviour, then that's ok =).

**Comment by vinivendra** [comment]
Hey guys, I'm coming with a fresh set of eyes to this issue. I think we should take a moment to remember that we should design the language we want, not the one that makes sense based on our own preconceptions of what languages are.

I'm saying this because I think we've been taking this discussion in the wrong direction. Firstly, for me, it makes no sense (semantically) that `Nil` should be a superclass to all. That would mean that all objects are `Nil`s, which is just weird. It would make more sense if `Nil` were somehow a subclass to all, but that would be a whole other can of worms.

In my opinion, `nil` should just be a value. It can be cast or assigned to any type, and it implements no functions, like we said. As an exception, I would allow for == to be implemented, only because it seems necessary to me to offer a way of checking whether or not a value is `nil`. This would also mean that the pattern matching cases would throw an error, as maybe they should - unless we check for `nil` as the first `case`, any messages sent to `nil` should throw errors, even `is`.

**Comment by vinivendra** [comment]
Also, I think we should try to remember our goals and principles in big decisions like these. We wanted to create a language that was different from the others because it had great performance, yet it was clear - which, we all agree, is something that doesn't exist today. This is reflected in our principles: clarity, simplicity, performance and standardization,

I mention this because I'd like to argue against having any kind of dynamic dispatch - that is, if it means having performance overhead, like you mentioned. If we want a language that may be used to build operating systems, graphics applications and scientific algorithms, it has to go fast - and having an indirection on every function call can be especially damaging in tight loops. It really does go against our principles.

Personally, I'd love to see other features in this language such as polymorphism, protocols (aka interfaces), etc. However, if that means sacrificing performance, I think the principles themselves answer those questions for us.

**Comment by igorbonadio** [comment]
@vinivendra , but what is the type of nil?

I dont see a problem if Nil is a super class... I learned that the class architecture doesnt need to be equal to the real world architecture... The main objective of OO is share behavior.

It makes sense to me, because of the type system. When you send a message to nil it breaks beacuse the superclass is very thin an just understand `==`. Its children implements `==` and other messages.

But if we found a solution that is closer to the real world... it is better.

**Comment by igorbonadio** [comment]
Performance is the last law... If we will cut off polymorphism I think positron shouldnt be OO... Or it will be equals C++ without virtual functions... which I think it's not useful.

And as I said, this kind of dynamic dispatch is really cheap.

**Comment by igorbonadio** [comment]
Ah, I think we should discuss polymorphism, dynamic dispatch, etc in other issue...

**Comment by igorbonadio** [comment]
I was thinking and I think I have other solution and you can see if it is better...

`nil` is not a value. `nil` is a language construct which means that some variable is not initialized.

But, in this case, `()` is different of `nil`... as `()` is a value and `nil` is not. 

Note that I use the word "different" not `!=`... if we want, we can implement `().is_nil()` to return `true`.

But they are different, because `()` is a value of type `()` (empty tuple) and `nil` is not a value.

It implies in other things, for example:

``` ruby
var x: Int = nil
x == nil # wrong... nil is not a value!
x.is_nil() # right: method call (it can be implemented in Object)
```

In other words, `nil` can be used only to empty a variable.

For example:

`if` is not an expression... but `if a == b then blah else blah2 end` is.

The same way, `nil` is not an expression, but `variable = nil` is.

Sorry if it was so confusing...

**Comment by vinivendra** [comment]
I was reading my comments again and I think I sounded a bit harsh, sorry for that :)

My problem with treating nil as a superclass is that expressions like `var x: Integer = nil` don't make sense. If a variable with a normal type ever contains nil, that should be an error, since it should contain (in this case) either an Integer or a subclass of Integer, but never a superclass. Does that make sense?

I see no problem in implementing an `is_nil()` function instead of ==, though I'm don't really like snake case. Anyway, I don't think `nil` should be different than `()`. I don't even mean to imply how they should respond to equality tests, I think they should be the same thing. If nil indicates that a variable has not been initialized, I don't think it's fair to say that a variable that has received `()` has been properly initialized:

``` ruby
let x = print("potato")
```

To me, x still doesn't have a valid value (print obviously returned `()`). Also, it seems to me that this approach doesn't bring any objective advantages: `()` would still throw an error whenever someone tried to call a method on it, only it couldn't be used with our `~>` try operator, forcing explicit checks all over.

What do you think?

**Comment by igorbonadio** [comment]
You right. It will only work if we allow downcast from `nil` to other type. And so, if you send a message to `nil` it breaks the program. But I agree that it is not the best solution.

But I don't like `() == nil` because, for me, both have different types. I think it is strange...

I think that, in your example, `x` is of type `()` (empty tuple). And the following code should not compile:

``` ruby
let num: Int = print("potato")
```

as `num` is of type `Int` and `print` is of type `()`. This cast is weirder than the downcast of Nil.

**Comment by igorbonadio** [comment]
I think the best option is set `nil` as a language construct. It does not have a type (or value). It just means that a variable is not initialized. Or that you are erasing a variable.

**Comment by vinivendra** [comment]
Ok, I'm fine with that. But then, in what case would the `()` value appear? I thought it wa just for when functions didn't return anything, but it does make much more sense to disallow that.

**Comment by igorbonadio** [comment]
In this case `nil` is not a value (an it doesn't have a type), so `()` is used when you need to return a value that means nothing.

As (I think) functions should always return a value, functions that return nothing should return `()`.

**Comment by vinivendra** [comment]
I get that, but in what situation would that value matter if, as the print example shows, we can never assign it to variable?

I also don't get this: what are the advantages of separating both? So far, as far as I can tell, they behave exactly the same way. We're only separating them because they show up from different scenarios, or because they mean slightly different things. But if they do the same thing, what's the point other than philosophy? To me, the complexity presented to the programmer (for having to reason with two "nothings" instead of one) seems to outweigh our will to stand by this philosophy.

**Comment by igorbonadio** [comment]
I think the `nil` keyword is confusing everything... because `nil` is not a value and it is used as a rhs of a attribution.

I will introduce a new syntax to try to solve it:

```
var x: Int = 1
x.next() # => 2
```

The code above is ok, right? It means: "there is a variable named `x` that stores values of type `Int` and its content is `1`. After that I send to x, which stores `1`, the message `next`. The result is 2"

Now, consider:

``` ruby
var x: Int
x.next() # => segmentation fault
```

This code means: "there is a variable named `x` that stores values of type `Int`. After that I send to x, which stores some trash, the message `next`. The result is unpredictable (something like a segmentation fault)"

So I ask: what is the value of `x`? I don't know...

How can we set an "trash" value to a value that stores a good value? I think we shouldn't use `nil`... My suggestion is:

``` ruby
var x: Int = 1
x.next() # => 2
var x: Int
x.next() # => segmentation fault
```

In this case: "there is a variable named `x` that stores values of type `Int` and its content is `1`. After that I send to x, which stores `1`, the message `next`. The result is 2. After that, I shadow the previous declaration of `x` with a new one. So now, there is a variable named `x` that stores values of type `Int`. After that I send to x, which stores some trash, the message `next`. The result is unpredictable (something like a segmentation fault)"

Other solution could be add a `clear` keyword:

``` ruby
var x: Int = 1
x.next() # => 2
clear x
x.next() # => segmentation fault
```

I think it solves the `nil` problem... that for me was: `nil` cant be a value of a type T. Because we cant cast T to any type (without doing something ugly). That's why I dont like `nil == ()`. `()` has a type. And if `nil` is equal to `()` so we have to do something ugly to cast it to other type.

Now, lets talk about functions:

I think functions should always return a tuple. And I think we all agree with it. So, if a function retuns 2 values, it returns a 2-tuple. If a functions returns 10 values, it returns 10-tuple. If a function returns 0 values, it returns a 0-tuple.

Ok, a `()` (0-tuple, empty tuple)  seems useless... but I think it makes functions declarations uniform.

To close my comment:

summary about nil:
- lets remove `nil` from the language
- uninitialized variables are variables that we define and don't set some value
- if you need to uninitialize a variable, so you redeclare it (shadowing the previous declaration) or use the clear keyword

summary about functions
- functions always return a tuple

**Comment by Kazuo256** [comment]
So, basically, the `clear` keyword is the same as `delete` from C++, except that it actually marks the variable as "trash value". Seems good.

**Comment by vinivendra** [comment]
I like it, not having nil does seem to make everything a lot simpler!

But then, I would argue not to have an explicit 0-tuple either: we don't allow assignments in which the rhs is a 0-tuple, and we make it so we never have to worry about it. Other than to match the philosophy that functions always return tuples, the 0-tuple never actually shows up in the code. What do you think?

**Comment by igorbonadio** [comment]
My problem with it is that, if a function may not return a value, so function calls are not expressions... And I think (almost) everything should be an expression.

**Comment by vinivendra** [comment]
I see. Now that i think of it, there are probably a few expressions that might have to be accounted for.

In this case, would we maintain the `try` operator? And should it work with both empty tuples and "trash value"?

**Comment by igorbonadio** [comment]
I think the try operator should work with abnormal execution:

``` ruby
var x: Int
x.next() ~> 0
```

In the case of `()`:

``` ruby
var y = () # y is of type ()
y.next() ~> 0 # it won't compile... as the type () doesn't define the method `next`
```

**Comment by vinivendra** [comment]
Ok, this sounds good. So:
- Variables that aren't initialized don't have any specific value.
- Functions with no return value return an empty tuple.
- The try operator safeguards against abnormal execution, which means saving programs that send messages to uninitialized variables.
- The compiler must have enough safeguards so that sending a method to an object that doesn't respond to it may never happen, thus making the try operator not applicable.

**Comment by igorbonadio** [comment]
Yes. I agree. What do you @renatocf @Kazuo256 think about it?

**Comment by Kazuo256** [comment]
Looks good.

**Comment by renatocf** [comment]
It seems nice. But, how will our variables act? If we don't have `nil`, they are neither pointers (in C/C++ way), nor references (in Java sense). I mean, in C++ it's perfectly possible to program without explicit use of pointers - and I have already read that non-standardized syntaxes like `->` and `.` could have been unified among static/stack/heap-allocated variables. But will we have pointers / references? Or can we survive without them (given Positron is a system programming language)? Maybe this is a topic for another issue (or a reason to reopen the issue about `nil`), but I think we have to discuss that.

**Comment by igorbonadio** [comment]
@renatocf , I think it is a very important subject and that we should discuss it in other issue.

So, I will close this one.
